### PR TITLE
Add multipart form samples for bff

### PR DESCRIPTION
**Pull request by SasinduDilshara**
Add code samples related to `rest` APIs with multi-part form data for ballerina backend to frontend(bff).

### CODE REVIEW

File: handle-mulipart-form-data/backend_server/.devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
+    "image": "ballerina/ballerina-devcontainer:2201.6.0",
+    "extensions": ["WSO2.ballerina"],
+}
```
Comment (LakshanWeerasinghe) [inline handle-mulipart-form-data/backend_server/.devcontainer.json]: we can remove this file

File: handle-mulipart-form-data/backend_server/server.bal [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    resource function post customers(http:Request request) returns http:Created|http:InternalServerError {
+        do {
+            mime:Entity[] bodyParts = check request.getBodyParts();
+            string registrationDataString = check bodyParts[0].getText();
+            json registrationDataJson = check registrationDataString.fromJsonString();
+            CustomerRegistrationData registrationData = check registrationDataJson.cloneWithType();
+            byte[] agreemntForm = check bodyParts[1].getByteArray();
+            byte[] image = check bodyParts[2].getByteArray();
+            string customerId = check registerCustomer(registrationData, agreemntForm, image);
+            return <http:Created>{body: {status: "Customer registered successfully.", customerId}};
```
Comment (LakshanWeerasinghe) [inline handle-mulipart-form-data/backend_server/server.bal]: Why do we need to send status message, if we are already send a `http:Created` status?

Comment (LakshanWeerasinghe) [reply handle-mulipart-form-data/backend_server/server.bal]: I feel like what we should return here is,
```ballerina
return <http:Created>{body: {result: {... registrationData, customerId}}};
```

**Review by LakshanWeerasinghe — CHANGES_REQUESTED** [review]

File: handle-mulipart-form-data/backend_server/server.bal [hunk 3]
```diff
@@ -0,0 +1,67 @@
+import ballerina/http;
+import ballerina/io;
+import ballerina/log;
+import ballerina/mime;
+import ballerina/random;
+
+public type CustomerRegistrationData record {|
```
Comment (LakshanWeerasinghe) [inline handle-mulipart-form-data/backend_server/server.bal]:
```suggestion 
type CustomerRegistrationData record {|
```

File: handle-mulipart-form-data/backend_server/server.bal [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    string firstName;
+    string lastName;
+    string address;
+    int dependents;
+|};
+
+final table<CustomerRegistrationData> customerTable = table [
```
Comment (LakshanWeerasinghe) [inline handle-mulipart-form-data/backend_server/server.bal]: This table should act as a data store. So this table values should hold the `customerId` field as well. Lets define a new `Customer` record and use the `Customer` record to create the table.
```ballerina
type Customer record {|
     *CustomerRegistrationData;
     readonly string customerId;
|};

final table<Customer> key(customerId) customers = table [];
```

File: handle-mulipart-form-data/backend_server/.devcontainer.json (hunk 1 above)
Comment (SasinduDilshara) [reply handle-mulipart-form-data/backend_server/.devcontainer.json]: Updated

File: handle-mulipart-form-data/backend_server/server.bal (hunk 2 above)
Comment (SasinduDilshara) [reply handle-mulipart-form-data/backend_server/server.bal]: Updated, set the key as `message` to be consistent with others

File: handle-mulipart-form-data/backend_server/server.bal (hunk 3 above)
Comment (SasinduDilshara) [reply handle-mulipart-form-data/backend_server/server.bal]: Updated

File: handle-mulipart-form-data/backend_server/server.bal (hunk 4 above)
Comment (SasinduDilshara) [reply handle-mulipart-form-data/backend_server/server.bal]: Yes, but this make sample bit complex ryt?
we only want to show the send and receive capabilities of multi-part data

File: handle-mulipart-form-data/backend_server/server.bal [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    resource function post customers(http:Request request) returns http:Created|http:InternalServerError {
+        do {
+            mime:Entity[] bodyParts = check request.getBodyParts();
+            string registrationDataString = check bodyParts[0].getText();
+            json registrationDataJson = check registrationDataString.fromJsonString();
+            CustomerRegistrationData registrationData = check registrationDataJson.cloneWithType();
+            byte[] agreemntForm = check bodyParts[1].getByteArray();
+            byte[] image = check bodyParts[2].getByteArray();
+            string customerId = check registerCustomer(registrationData, agreemntForm, image);
+            return <http:Created>{body: {message: "Customer registered successfully.", customerId}};
```
Comment (chathurace) [inline handle-mulipart-form-data/backend_server/server.bal]: I think it's better to store customerId in the table and also store agreementForm and image in the file system against the customerId (e.g. <customerId>_agreement.pdf, <customerId>_image.png). Then we can return the actual agreement in the /customers/<id>/agreement resource.

File: handle-mulipart-form-data/backend_server/server.bal [hunk 6]
```diff
@@ -0,0 +1,67 @@
+import ballerina/http;
+import ballerina/io;
+import ballerina/log;
+import ballerina/mime;
+import ballerina/random;
+
+type CustomerRegistrationData record {|
```
Comment (chathurace) [inline handle-mulipart-form-data/backend_server/server.bal]: We can rename this to CustomerData

File: handle-mulipart-form-data/backend_server/server.bal [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+}
+service /crm on new http:Listener(9090) {
+    resource function get customers() returns CustomerRegistrationData[] {
+        return customerTable.toArray();
+    }
+
+    resource function get customers/[string customerId]/agreement() returns byte[]|http:NotFound {
+        byte[]|error agreementForm = getAgreementForm(customerId);
+        if agreementForm is error {
+            return <http:NotFound>{body: {message: "Agreement form not found for the customer ID: " + customerId}};
```
Comment (chathurace) [inline handle-mulipart-form-data/backend_server/server.bal]: If the file read fails we cannot return NoFound error. Better to return a generic error such as 500.

File: handle-mulipart-form-data/backend_server/server.bal [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+            mime:Entity[] bodyParts = check request.getBodyParts();
+            string registrationDataString = check bodyParts[0].getText();
+            json registrationDataJson = check registrationDataString.fromJsonString();
+            CustomerRegistrationData registrationData = check registrationDataJson.cloneWithType();
+            byte[] agreemntForm = check bodyParts[1].getByteArray();
+            byte[] image = check bodyParts[2].getByteArray();
+            string customerId = check registerCustomer(registrationData, agreemntForm, image);
+            return <http:Created>{body: {message: "Customer registered successfully.", customerId}};
+        } on fail error e {
+            return <http:InternalServerError>{body: {message: string `Error while parsing the request: ${e.message()}`}};
```
Comment (chathurace) [inline handle-mulipart-form-data/backend_server/server.bal]: Let's change to "Error while processing the request"

File: handle-mulipart-form-data/backend_server/server.bal [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        }
+    }
+}
+
+function getAgreementForm(string id) returns byte[]|error {
+    log:printInfo("Agreement form requested for customer ID: " + id);
+    string filePath = getAgreementFormPath();
```
Comment (chathurace) [inline handle-mulipart-form-data/backend_server/server.bal]: Change to return the actual form.

File: handle-mulipart-form-data/backend_server/server.bal (hunk 4 above)
Comment (SasinduDilshara) [reply handle-mulipart-form-data/backend_server/server.bal]: Updated.

File: handle-mulipart-form-data/backend_server/server.bal (hunk 6 above)
Comment (SasinduDilshara) [reply handle-mulipart-form-data/backend_server/server.bal]: Updated.

File: handle-mulipart-form-data/backend_server/server.bal (hunk 7 above)
Comment (SasinduDilshara) [reply handle-mulipart-form-data/backend_server/server.bal]: Updated.

File: handle-mulipart-form-data/backend_server/server.bal (hunk 5 above)
Comment (SasinduDilshara) [reply handle-mulipart-form-data/backend_server/server.bal]: Updated.

File: handle-mulipart-form-data/backend_server/server.bal (hunk 8 above)
Comment (SasinduDilshara) [reply handle-mulipart-form-data/backend_server/server.bal]: Updated.

File: handle-mulipart-form-data/backend_server/server.bal (hunk 9 above)
Comment (SasinduDilshara) [reply handle-mulipart-form-data/backend_server/server.bal]: Updated.

File: handle-mulipart-form-data/frontend_server/src/pages/create-customer.js [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+      }
+    } catch (err) {
+      console.log(err)
+      setErrorMessage(err.data.response.data.message);
+      setOpen(true);
+    }
+  }
+
+  return (
+    // error != null ? <div>Something went wrong</div> :
```
Comment (sm1990) [inline handle-mulipart-form-data/frontend_server/src/pages/create-customer.js]: Remove unused code.

Comment (SasinduDilshara) [reply handle-mulipart-form-data/frontend_server/src/pages/create-customer.js]: Updated.

**Review by anupama-pathirage — APPROVED** [review]

**Review by sameerajayasoma — APPROVED** [review]
Reviewed the Ballerina code

**Review by LakshanWeerasinghe — APPROVED** [review]

**Pull request merged**